### ISSUE TITLE
Wrong results when using Having / distinct / aggregation function 

### ISSUE THREAD

**Issue by Alkaagr81** [pingcap/tidb#29368]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
CREATE TABLE t1 (a INTEGER);
INSERT INTO t1 VALUES (1), (2);
SELECT DISTINCT COUNT(DISTINCT t1.a) AS da FROM t1 JOIN t1 AS t2 ON t1.a = t2.a
GROUP BY t1.a HAVING COUNT(DISTINCT t1.a) = 1;
DROP TABLE t1;

```
### 2. What did you expect to see? (Required)
```sql
mysql> DROP TABLE t1;
Query OK, 0 rows affected (0.01 sec)

mysql> CREATE TABLE t1 (a INTEGER);
Query OK, 0 rows affected (0.01 sec)

mysql> INSERT INTO t1 VALUES (1), (2);
Query OK, 2 rows affected (0.00 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql> SELECT DISTINCT COUNT(DISTINCT t1.a) AS da FROM t1 JOIN t1 AS t2 ON t1.a = t2.a
    -> GROUP BY t1.a HAVING COUNT(DISTINCT t1.a) = 1;
Empty set (0.00 sec)

mysql> DROP TABLE t1;
Query OK, 0 rows affected (0.01 sec)

```
### 3. What did you see instead (Required)
```sql
mysql> CREATE TABLE t1 (a INTEGER);
Query OK, 0 rows affected (0.07 sec)

mysql> INSERT INTO t1 VALUES (1), (2);
Query OK, 2 rows affected (0.01 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql> SELECT DISTINCT COUNT(DISTINCT t1.a) AS da FROM t1 JOIN t1 AS t2 ON t1.a = t2.a
    -> GROUP BY t1.a HAVING COUNT(DISTINCT t1.a) = 1;
+----+
| da |
+----+
|  1 |
+----+
1 row in set (0.00 sec)

mysql> DROP TABLE t1;
Query OK, 0 rows affected (0.25 sec)

```
### 4. What is your TiDB version? (Required)

```sql
| Release Version: v5.2.2
Edition: Community
Git Commit Hash: da1c21fd45a4ea5900ac16d2f4a248143f378d18
Git Branch: heads/refs/tags/v5.2.2
UTC Build Time: 2021-10-20 06:03:45
GoVersion: go1.16.5
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
```

**Comment by guo-shaoge** [comment]
IMHO, TiDB’s result is correct. Also following is pg13'result. And I report a [mysql bug](https://bugs.mysql.com/bug.php?id=105437&thanks=4).

    postgres=# CREATE TABLE t1 (a INTEGER);
    RT INTO t1 VALUES (1), (2);CREATE TABLE
    postgres=# INSERT INTO t1 VALUES (1), (2);
    INSERT 0 2
    postgres=# select * from t1;
     a
    ---
     1
     2
    (2 rows)
    
    postgres=# SELECT DISTINCT COUNT(DISTINCT t1.a) AS da FROM t1 JOIN t1 AS t2 ON t1.a = t2.a
    postgres-# GROUP BY t1.a HAVING COUNT(DISTINCT t1.a) = 1;
     da
    ----
      1
    (1 row)
    
    postgres=# select version();
                                                          version
    -------------------------------------------------------------------------------------------------------------------
     PostgreSQL 13.2 on x86_64-pc-linux-gnu, compiled by clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final), 64-bit
    (1 row)

**Comment by guo-shaoge** [comment]
In MySQL 8.0.27, the result is same as TiDB.

    MySQL [test]> CREATE TABLE t1 (a INTEGER);
    Query OK, 0 rows affected (0.010 sec)
    MySQL [test]> INSERT INTO t1 VALUES (1), (2);
    Query OK, 2 rows affected (0.008 sec)
    Records: 2  Duplicates: 0  Warnings: 0
    MySQL [test]> SELECT DISTINCT COUNT(DISTINCT t1.a) AS da FROM t1 JOIN t1 AS t2 ON t1.a = t2.a GROUP BY t1.a HAVING COUNT(DISTINCT t1.a) = 1;
    +----+
    | da |
    +----+
    |  1 |
    +----+
    1 row in set (0.001 sec)
    MySQL [test]> select version();
    +-----------+
    | version() |
    +-----------+
    | 8.0.27    |
    +-----------+
    1 row in set (0.000 sec)

**Comment by winoros** [comment]
Since it's not TiDB's bug. We move down the severity. And remove the bug label.

**Comment by morgo** [comment]
Confirmed it was a MySQL bug. Fixed in 8.0.27:
```sql
mysql [localhost:8027] {msandbox} (test) > SELECT DISTINCT COUNT(DISTINCT t1.a) AS da FROM t1 JOIN t1 AS t2 ON t1.a = t2.a GROUP BY t1.a HAVING COUNT(DISTINCT t1.a) = 1;
+----+
| da |
+----+
|  1 |
+----+
1 row in set (0.00 sec)
```
